src/base/memory_monitor_sr.cpp

```diff
@@ -92,6 +96,72 @@ namespace cubmem
 
     return filecopy + ':' + std::to_string (line);
   }
+
+  void memory_monitor::add_stat (char *ptr, const size_t size, const char *file, const int line)
+  {
+    std::string tag_name;
+    MMON_METAINFO metainfo;
+
+    // size should not be 0 because of MMON_ALLOC_META_SIZE
+    assert (size > 0);
+
+    metainfo.allocated_size = (uint64_t) size;
+    m_total_mem_usage += metainfo.allocated_size;
+
+    tag_name = make_tag_name (file, line);
+
+    std::unique_lock <std::mutex> tag_map_lock (m_tag_map_mutex);
+    auto search = m_tag_map.find (tag_name);
+    if (search != m_tag_map.end ())
+      {
+	metainfo.tag_id = search->second;
+	m_stat_map[metainfo.tag_id] += metainfo.allocated_size;
+      }
+    else
+      {
+	metainfo.tag_id = m_tag_map.size ();
+	// tag_id is start with 0
+	m_tag_map.emplace (tag_name, metainfo.tag_id);
+	m_stat_map.emplace (metainfo.tag_id, metainfo.allocated_size);
+      }
+    tag_map_lock.unlock ();
+
+    // put meta info into the allocated chunk
+    char *meta_ptr = ptr + metainfo.allocated_size - MMON_ALLOC_META_SIZE;
```

Comment:
I'd extract this as a static private function like `get_meta_ptr (ptr, user_size)` because the same calculation which is used across functions and must be the same. It's safer.

Reply:
I agree. I add it.